### PR TITLE
Drivers: DMIC: Updates to enable runtime instantiate and remove DAI

### PR DESCRIPTION
**Pull request by singalsu**
This PR version is the first that passed my tests with first DMIC DAI set to 48 kHz and second DAT set to 8/16/32/96 kHz with the same binary firmware image (no Kconfig changes, DMIC full set used).

I'm still keeping this as draft to see CI tests results.

### CODE REVIEW

**Comment by lgirdwood** [comment]
@singalsu can you include the benefits in the PR/commit messages. i.e. why we need this invasive change.

**Comment by singalsu** [comment]
This version contains a proper FIR filters fix.

@lgirdwood Sure, I'll improve the comments and commit messages into next version. I'm still testing that this works with all the rates combinations. I'm still not sure everything is possible if the proactive DAIs initialization is removed.

**Comment by singalsu** [comment]
@ranj063 @lgirdwood Without proactive initialize of DAIs pretty much anything else than 48 kHz and 16 kHz rates combination will fail start if higher rate capture is started just after running lower rate capture. If the proactive init is not done something like this is needed:
- Optional force of proactive initialization
- Optional force of DMIC clock divider value and CIC decimation ratio from topology. Some reserved fields in IPC could pass these. Currently the topology writer do not need to know such details. The clock divider could be already now forced to a value with min==max clock parameter but it can't solve all cases.
-  Other DMIC DAI rate information to DMIC IPC. This would allow the driver to select the above parameters to be compatible with a not yet received request.

What do you think?

**Comment by plbossart** [comment]
@singalsu what does 'proactive initialization' mean? You lost me in the first sentence :-)

**Comment by ranj063** [comment]
> @singalsu what does 'proactive initialization' mean? You lost me in the first sentence :-)

@plbossart before this PR, I was running into trouble with starting the second DMIC pipeline as the DAI config IPC for the second DMIC DAI while the first one was active. 
So, I added a workaround in the driver to always create the DMIC DAIs and configure them after resume even if there werent any active DMIC pipelines during suspend. That's what Seppo is referring to as proactive init.

**Comment by plbossart** [comment]
> > @singalsu what does 'proactive initialization' mean? You lost me in the first sentence :-)
> 
> @plbossart before this PR, I was running into trouble with starting the second DMIC pipeline as the DAI config IPC for the second DMIC DAI while the first one was active ran into a DSP panic.
> So, I added a workaround in the driver to always create the DMIC DAIs and configure them after resume even if there werent any active DMIC pipelines during suspend. That's what Seppo is referring to as proactive init.

Ah, makes sense now. Thanks Ranjani for helping me connect the dots.

**Comment by singalsu** [comment]
Here's my test round with 48 kHz and another common sample rate

- 8 & 48 kHz, fails if 48 kHz starts during 8 kHz capture, could be made work with remove of decimation factor 10
- 16 & 48 kHz, OK with this PR
- 32 & 48 kHz, can be made OK with slight reorder or decimation factors preference
- 96 & 48 kHz, can be made by forcing min. clock to 3.8 MHz

If there's no information about other DAI rate there would be need with "make menuconfig" disable decimation factors those cause a "blocked route" to setup the other DAI. With binary Intel signed firmwares only it's not a good way.

**Comment by singalsu** [comment]
New version! The earlier mentioned problems were solved. I also improved the commit messages to describe the benefit and not just the change.

**Comment by lgirdwood** [comment]
@singalsu so does the new version solve all the reported issues or are there still some restrictions on ordering ?

**Comment by singalsu** [comment]
@lgirdwood @ranj063 This version survived all my tests with FIFOA set to 48 kHz and FIFOB set to 8/16/32/48/96 kHz.  The test script that I use does two overlapping captures, e.g. first capture 48 kHz then 16 kHz, first 16 kHz, then 48 kHz. etc. We use today only 48 kHz and 16 kHz in topologies. There was a study done per customer ask about 96 kHz but the requirement didn't happen. It remains available if it becomes a requirement. It needs only a few 48000 -> 96000 edits in topology m4.

But I'm pretty sure there could be fails with entire matrix of 8/16/32/48/96 and 8/16/32/48/96. Also I didn't test other possible rates like 12, 24, 64 kHz. The proactive DAI initialize by Linux kernel is the only way to de-risk blocked paths in setting up decimation filters in HW for all rates.

Also the cAVS firmware for Windows supports only 16 kHz and 48 kHz, nothing else.

File: src/drivers/intel/dmic.c [hunk 1]
```diff
@@ -579,9 +578,19 @@ static int select_mode(struct dai *dai,
...
+	 */
+	count = 0;
+	for (n = 0; fir_list[n]; n++) {
+		count = find_equal_int16(idx, mfir, fir_list[n]->decim_factor,
+					 modes->num_of_modes, 0);
+		if (count > 0)
+			break;
+	}
+
+	assert(count); /* Should not happen. */
```
Comment (lgirdwood) [inline src/drivers/intel/dmic.c:592]: Can we have an trace error here instead and return an error to make this survivavble.

**Review by lgirdwood — CHANGES_REQUESTED** [review]

File: src/drivers/intel/dmic.c (hunk 1 above)
Comment (singalsu) [reply src/drivers/intel/dmic.c]: Sure, I changed this just to error trace and -EINVAL return. Only a developer doing changes to to decimation ratios selection should be able to get this error. I tried to force this to happen, seems like the DMA trace from topology load phase does not show this. But there's clean error in dmesg at least.

**Comment by lgirdwood** [comment]
@singalsu intended for merge now ? It's still draft.

**Comment by singalsu** [comment]
@ranj063 @lgirdwood How would you like to proceed? The CI seems to be passed and I feel this is good to go. Maybe it would not hurt to merge this early so there would be plenty of time for testing and make sure we don't get regressions before next release.

**Comment by lgirdwood** [comment]
@singalsu you need to change the PR from draft status if you think it's ready to merge.

**Comment by singalsu** [comment]
I have created this PR incorrectly from a SOF repo branch while I had my git incorrectly configured. I'll close this and create new PR. I think it's up to @ranj063 to decide if we should use this change or not.

**Comment by lgirdwood** [comment]
@singalsu safe to delete the branch ?

**Pull request closed without merging**